src/server/frontend_xwayland/xwayland_surface_role.cpp

```diff
@@ -125,7 +119,13 @@ void mf::XWaylandSurfaceRole::commit(WlSurfaceState const& state)
         BOOST_THROW_EXCEPTION(std::runtime_error("Got XWaylandSurfaceRole::commit() when the role had no surface"));
     }
 
-    wl_surface->commit(state);
+    // HACK: Some applications might scale up while users don't want them to
+    // `--x11-scale` helps with this by scaling down the application according
+    // to user preference.
+    auto state_copy{state};
+    state_copy.scale = scale;
```

Comment:
Oops! C++ will silently convert from `float` to `optional<int>`, meaning this only works for integer scale. (If you try `--x11-scale=0.5` Mir will immediately crash when the first X11 app starts, because this truncates `0.5f32` to `0i32`, and we later divide by `scale`)